### ISSUE TITLE
[Live Share: Guest initiated debugging] You don't have an extension for debugging .....

### ISSUE THREAD

**Issue by jramsay** [microsoft/vscode#132595]
Does this issue occur when all extensions are disabled?: Yes/No

Version: 1.60.0 (user setup)
Commit: e7d7e9a9348e6a8cc8c03f877d39cb72e5dfb1ff
Date: 2021-09-01T10:41:52.311Z
Electron: 13.1.8
Chrome: 91.0.4472.164
Node.js: 14.16.0
V8: 9.1.269.39-electron.0
OS: Windows_NT x64 10.0.18363

Steps to Reproduce:

1. Live Share host starts a collaboration session
2. Live Share guest joins session
3. Live Share guest attempts to initiate debugging

Hi @isidorn: there are a bunch of similar issues that are already closed. I've seen your responses that adding the debug adapter's supported debugger "languages"  to the package.json resolves [the issue](https://github.com/microsoft/vscode/issues/123881).

The live share guest's debug adapter should support all languages that the host can debug though. So we would need to add every possible language so that the debug request could be forwarded to the host. If the guest is in a launch.json file they see this though: 

![image](https://user-images.githubusercontent.com/8161247/132421047-dd68b78c-2622-4b6b-b524-3a37c84f449a.png)

Is there a wildcard option for all languages or a way to disable this new restriction for the Live Share guest?

**Comment by isidorn** [comment]
@jramsay I see that live share is a special case since the guest can start any debug session even though a `launch.json` is not present.
Having a wildcard for the `languages` is maybe too general as it might have other implications.
I think we arleady treat live share specially because it has a `type` of `*`. I suggest to simply piggy back on that flag.

So if the debugger has a `*` as a `type` the dialog should not appear. Code pointer. PR's are welcome
https://github.com/microsoft/vscode/blob/main/src/vs/workbench/contrib/debug/browser/debugAdapterManager.ts#L304

Though I leave the final decision to @weinand and @roblourens since they own debug now.

**Comment by weinand** [comment]
I agree with Isidorn that a local fix to prevent the dialog is the preferred low-risk fix for this problem.

The check could be [in this line](https://github.com/microsoft/vscode/blob/08370d00b48266a911172d38173c4c8ef8c73e46/src/vs/workbench/contrib/debug/browser/debugAdapterManager.ts#L270)

And the code would basically check if a debugger is registered for the debug type `*` and then return it.

Now let's analyse the consequences for users that have the LiveShare debugger installed:

if users want to debug a file for which they do not have a installed debugger (and no launch.json exists), they will no longer get the Marketplace dialog but they will end up in the LiveShare debugger instead.

@jramsay Is this the expected behavior?

Or should we ask users first (with another dialog) whether they want to use LiveShare debugging before the Marketplace dialog is shown? The downside of this approach is that more LiveShare knowledge gets into our code...

**Comment by roblourens** [comment]
I'm curious how this works in other cases, as a guest do I get intellisense via LiveShare for Python files even if I don't have the Python extension installed?

What if the host doesn't have the debugger installed either?

> if users want to debug a file for which they do not have a installed debugger (and no launch.json exists), they will no longer get the Marketplace dialog but they will end up in the LiveShare debugger instead.

And this is while not in a liveshare session? I would want to avoid things that impact the experience when the liveshare extension is installed but you are not in a liveshare session.

**Comment by jramsay** [comment]
Sorry for the delay! (OOF) Live Share currently registers two debuggers in the extension's package.json:

- "type": "vslsJoin"
- "type": "vslsShare"

When the Live Share guest joins a collaboration session we initialize our JoinDebugManager and this registers a debugConfigurationProvider:

vscode.debug.registerDebugConfigurationProvider('*', this);

This is used to intercept all possible debug types but only for the duration of the collaboration session.

If the host does something similar on sharing a collaboration session.

I debugged the line you mentioned in the debugAdapterManager. I _don't think_ this will work though as the debugger itself is of type 'vslsJoin'. Instead we'd want to skip the marketplace dialog only if there is an IDebugConfigurationProvider that is registered to intercept all types. This would be the case when the Live Share guest is actually in a collaboration session.

It would involve 2 changes:

1. Add the following to Live Share's joinDebugManager (vscode.DebugConfigurationProvider): 

   public async provideDebugConfigurations(folder: vscode.WorkspaceFolder | undefined, token?: vscode.CancellationToken): Promise<vscode.DebugConfiguration[]> {
        return [];
    }

2. In debugService.ts' CreateSession, when the configurationManager has a debugConfigurationProviders of both type '*' && 'vslsJoin' we'd set the return type to be 'vslsJoin'

![image](https://user-images.githubusercontent.com/8161247/134997036-7c081514-9e87-48ee-ad5d-dd54a053d655.png)

When I do this, when the Live Share guest initiates debugging an empty launch.json is created if it doesn't already exist:

![image](https://user-images.githubusercontent.com/8161247/134997454-1e3a937a-9478-42d3-b28f-21a809210510.png)

The guest has to manually edit this launch.json to add a debug configuration and then they are able to start debugging:

ex: 
![image](https://user-images.githubusercontent.com/8161247/134997624-5e7f04fb-e273-4d08-a1e9-5e698643dc8b.png)


This is acceptable but it is worth noting that previously when the launch,json wasn't present - it was generated but the Live Share guest didn't have to manually add the debug configuration. I can't remember if it was already filled out or if they were still able to launch the debug session with the empty launch.json - but recall this step wasn't necessary. Probably not an issue though as it seems straight-forward that you'd need to add the correct configuration to start debugging.

I'll send out a PR for this change. Although I'm sure setting the 'vslsJoin' type isn't ideal :)

**Comment by roblourens** [comment]
Ping, is there anything still to be done here?
